### ISSUE TITLE
Send token in Authorization header (apps_connections_open not working)

### ISSUE THREAD

**Issue by vramaiah** [slack-ruby/slack-ruby-client#363]
When I try to open a web socket connection using the apps_connections_open method, I receive a `not_allowed_token_type` error - 

```
client = Slack::Web::Client.new(token: 'bot_token')
client.apps_connections_open
Slack::Web::Api::Errors::NotAllowedTokenType: not_allowed_token_type
```

Creating the client with the app_token results in an invalid_auth -
```
client = Slack::Web::Client.new(token: 'app_level_token')
client.apps_connections_open
Slack::Web::Api::Errors::InvalidAuth: invalid_auth
```

Am I doing this incorrectly?

**Comment by dblock** [comment]
Usually `not_allowed_token_type` means that you have a legacy app trying to call an API that only works with granular permissions. Is it the case? [This](https://code.dblock.org/2020/11/30/migrating-classic-slack-ruby-bots-to-granular-permissions.html) might be helpful.

**Comment by vramaiah** [comment]
No. This is not a Legacy App. I was trying out the socket mode implementation which requires to POST to the `apps.connections.open` endpoint - https://api.slack.com/apis/connections/socket-implement. I corrected my message above. Using the bot_token gives a not_allowed_token_type since the `apps.connections.open` requires an app-level-token, but using the app_level_token returns an invalid_auth.

**Comment by dblock** [comment]
I don't know why this happens, I suggest opening a ticket with Slack and they will clear it up quickly.

**Comment by vramaiah** [comment]
I don't think it's an issue with Slack actually. Running curl against that endpoint works fine. I think the slack-ruby-client is using the bot_token to hit the apps.connections.open endpoint when it should actually send the app_token as an authorization bearer token in the header. I wonder if there's some way to force that for the methods that require the `app_token` within this client gem

**Comment by dblock** [comment]
Strange, the second example should have worked as long as the token is correct. The [apps_connections_open](https://github.com/slack-ruby/slack-ruby-client/blob/cd39ecd5c1cf58bea2fc7cb970a9c4522f8f9a5a/lib/slack/web/api/endpoints/apps_connections.rb#L15) method is just a wrapper that does a POST and uses a token [here](https://github.com/slack-ruby/slack-ruby-client/blob/cd39ecd5c1cf58bea2fc7cb970a9c4522f8f9a5a/lib/slack/web/faraday/request.rb#L25). 

If it works with `curl`, debug it all the way down through the method calls?

**Comment by vramaiah** [comment]
I got it to work with a small change to the `request` method [here](https://github.com/slack-ruby/slack-ruby-client/blob/cd39ecd5c1cf58bea2fc7cb970a9c4522f8f9a5a/lib/slack/web/faraday/request.rb)

The documentation for [Socket Mode Implementation](https://api.slack.com/apis/connections/socket-implement) specifies that the app_level_token needs to be sent in the `Authorization` header and not as a parameter. 

The [Slack post request ](https://github.com/slack-ruby/slack-ruby-client/blob/cd39ecd5c1cf58bea2fc7cb970a9c4522f8f9a5a/lib/slack/web/faraday/request.rb#L25) seems to send the token always in the body of the request as opposed to the header. Changing the request method to add the token to the Authorization header instead of the body seems to fix the problem.

This is what I did -
```
def request(method, path, options)
  response = connection.send(method) do |request|
    case method
    when :get, :delete
      request.url(path, options)
    when :post, :put
      request.path = path
      request.body = options unless options.empty?
    end
    request.headers['Authorization'] = "Bearer #{token}"
    request.options.merge!(options.delete(:request)) if options.key?(:request)
  end
  response.body
end
```

I haven't tested this completely but was able to both open a socket connection with an app token and also post messages with a bot_token. Sending it in the Authorization HTTP header seems to be the preferred manner per Slack as well- https://api.slack.com/web#slack-web-api__authentication

Hope this gives some insight and is helpful.

**Comment by dblock** [comment]
Oh that's very interesting! I suppose they deprecated the behavior for new APIs, we should fix this. The docs now say:

_We prefer tokens to be sent in the Authorization HTTP header of your outbound requests. However, you may also pass tokens in all Web API calls as a POST body parameter called token. Tokens cannot be sent as a query parameter._

@vramaiah Want to PR this change with additional specs that use `apps_connections_open`? You will need to re-run/edit a bunch of VCR specs, and double check that everything is working.

**Comment by vramaiah** [comment]
Hi @dblock I can take a stab at it. Will take me a bit though. Will keep you updated.

**Comment by raemillz** [comment]
Hi, I also ran into this issue and was wondering if there are any updates on @vramaiah's solution?

**Comment by vramaiah** [comment]
Not yet. I haven't had a chance to implement the fix yet. Can you maybe try this as a fix and if it solves the issue, we'll have validated the fix - https://github.com/slack-ruby/slack-ruby-client/issues/363#issuecomment-777187101

**Comment by raemillz** [comment]
@vramaiah that seems to solve the problem, nice digging!

**Comment by dblock** [comment]
Now maybe one of you can take the time to turn it into a proper PR? ;)

**Comment by jmanian** [comment]
I just ran into this issue while trying to use [`apps.event.authorizations.list`](https://api.slack.com/methods/apps.event.authorizations.list) with an app-level token.

I messaged Slack support and learned that app-level tokens must always be passed in a header, not as a parameter. At the moment this applies to just these two methods:
1. `apps.connections.open`
2. `apps.event.authorizations.list`

The proposed solution should fix both methods and any other methods that use app-level tokens going forward.
